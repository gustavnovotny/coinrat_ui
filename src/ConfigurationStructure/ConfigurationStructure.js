import {extendObservable} from "mobx"

const TYPE_STRING = 'string'
const TYPE_INTEGER = 'int'
const TYPE_DECIMAL = 'Decimal'

const POSSIBLE_TYPES = TYPE_STRING | TYPE_INTEGER | TYPE_DECIMAL

class ConfigurationDirective {
  key: string
  type: POSSIBLE_TYPES
  title: string
  _value: string | number
  defaultValue: string | number
  unit: ?string
  isRequired: boolean
  isDisabled: boolean

  constructor(
    key: string,
    type: POSSIBLE_TYPES,
    title: string,
    defaultValue: string | number,
    unit: ?string,
    isRequired: boolean,
    isDisabled: boolean
  ) {
    this.key = key
    this.type = type
    this.title = title

    extendObservable(this, {
      _value: defaultValue
    })

    this.defaultValue = defaultValue
    this.unit = unit
    this.isRequired = isRequired
    this.isDisabled = isDisabled
  }

  set value(value) {
    if (this.type === TYPE_INTEGER) {
      value = Number(value)
    }
    this._value = value
  }

  get value() {
    return this._value
  }

  resetToDefault = () => {
    this._value = this.defaultValue
  }
}

class ConfigurationStructure {
  configuration: Array<ConfigurationDirective>

  constructor(configuration: Array<ConfigurationDirective>) {
    extendObservable(this, {
      configuration: configuration,
    })
  }
}

function isDisabledByFieldName(key: string): boolean {
  return ['mocked_market_name'].includes(key)
}

const createConfigurationStructureFromRawData = (data) => {
  const configuration = Object.keys(data).map((key: string) => {
    const rawDirective = data[key]

    return new ConfigurationDirective(
      key,
      rawDirective.type.replace('?', ''),
      rawDirective.title,
      rawDirective.default,
      rawDirective.unit || null,
      !rawDirective.type.startsWith('?'),
      isDisabledByFieldName(key)
    )
  })

  return new ConfigurationStructure(configuration)
}

export {
  ConfigurationStructure,
  ConfigurationDirective,
  createConfigurationStructureFromRawData,
  POSSIBLE_TYPES,
  TYPE_STRING,
  TYPE_INTEGER,
  TYPE_DECIMAL,
}
