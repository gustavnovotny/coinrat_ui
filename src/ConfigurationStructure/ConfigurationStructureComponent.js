import React from "react"
import {Popover2} from "@blueprintjs/labs"
import {Button, Tooltip, Position} from "@blueprintjs/core"
import {
  ConfigurationDirective, ConfigurationStructure,
  TYPE_INTEGER, TYPE_STRING, TYPE_TIME_DELTA
} from "./ConfigurationStructure"
import {Box, Flex} from "reflexbox"
import FormItemComponent from "../Form/FormItemComponent"

type Props = {
  configurationStructure: ConfigurationStructure,
}

const ConfigurationStructureComponent = ({configurationStructure}: Props) => {

  const configurationMapFunction = (directive: ConfigurationDirective, key: number) => {
    let element = ''
    if (directive.type === TYPE_INTEGER) {
      element = <input className="pt-input" type="number"/>

    } else if (directive.type === TYPE_STRING) {
      element = <input className="pt-input" type="text"/>

    } else if (directive.type === TYPE_TIME_DELTA) {
      element = <input className="pt-input" type="number"/>
    }

    return <Box key={key}><FormItemComponent label={directive.title} element={element}/></Box>
  }

  const content = <Flex column style={{padding: 15 + 'px'}}>
    {configurationStructure.configuration.map(configurationMapFunction)}
  </Flex>

  const target = <Tooltip content="You can change settings of a strategy." position={Position.BOTTOM}>
    <Button
      style={{marginTop: 1 + 'px', marginLeft: 7 + 'px'}}
      className="pt-intent-warningÄ›"
      iconName="pt-icon-settings"
    />
  </Tooltip>

  return <Popover2 content={content} target={target}/>
}

export default ConfigurationStructureComponent
