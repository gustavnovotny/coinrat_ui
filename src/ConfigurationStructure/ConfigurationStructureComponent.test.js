// @flow
import React from 'react'
import renderer from 'react-test-renderer'
import ConfigurationStructureComponent from "./ConfigurationStructureComponent"
import {ConfigurationDirective, ConfigurationStructure, TYPE_INTEGER} from "./ConfigurationStructure"

it.skip('renders correctly', () => {
  global.document = {} // Todo: solve ReferenceError: document is not defined

  const configurationStructure = new ConfigurationStructure([
    new ConfigurationDirective('foo', TYPE_INTEGER, 'Foo', 1, 'lumen', true, false, 'This  is Foo!'),
  ])

  const tree = renderer
    .create(
      <ConfigurationStructureComponent
        tooltip={"Here is some tooltip!"}
        noConfigurationDescription={"Lorem ipsum dolor sit et amet!"}
        configurationStructure={configurationStructure}
        onChange={() => undefined}
        onReset={() => undefined}
      />)
    .toJSON()
  expect(tree).toMatchSnapshot()
})
