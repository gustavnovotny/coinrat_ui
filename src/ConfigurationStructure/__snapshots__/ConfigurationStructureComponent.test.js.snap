// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<span>
  <div
    className="pt-popover-target"
    onClick={[Function]}
  >
    <span>
      <div
        className="pt-popover-target"
        onBlur={[Function]}
        onFocus={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        <button
          className="pt-button pt-icon-settings"
          disabled={undefined}
          onClick={undefined}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          style={
            Object {
              "marginTop": "1px",
            }
          }
          tabIndex={0}
          type="button"
        />
      </div>
    </span>
  </div>
</span>
`;
